README.md
# Diffbot API Perl Client

# Overview

This module provides a Perl client for the diffbot REST API. 

Details of the REST API can be found here : http://www.diffbot.com/. 

You will need a token from diffbot to use their service.

# Installation

The following are the environment prerequisites:
 
You will need _make_ installed in your environment.

    In debian based systems (e.g., ubuntu, mint, etc):
    sudo apt-get install build-essential

You will need to install some perl dependancies:

    cpan install LWP::UserAgent Data::Dump JSON Test::Deep Test::Exception Test::Fake::HTTPD Getopt::Long

Ok, now you can install the actual module after downloading and extracting:

    tar xzvf Diffbot-Client-0.01.tar.gz
    cd Diffbot-Client-0.01
    perl Makefile.PL
    make
    make test
    sudo make install

# Documentation

This module contains built in documentation that is browsable once the module is installed.

To view the documentation after installation via manpages:

    man Diffbot::Client

To view the documentation after installation via perldoc: 

    perldoc Diffbot::Client

# Usage

## Make Diffbot fetch and analyze content using the article, frontpage, product, and image API
    
Once you have received a token from The Diffbot, you begin using the client module by instantiatng it:

    use Diffbot::Client
    my $client = Diffbot::Client->new({ token => '123456789' });

If you need manipuate the WWW Client config you can provide your own instantiated LWP::UserAgent:

    use Diffbot::Client;
    use LWP::UserAgent;
    my $ua = LWP::UserAgent->new( agent => 'Custom UserAgent!!' );
    my $client = Diffbot::Client->new({ token => '123456789', ua => $ua });

Now all you have to do is begin calling sub routines to get results!

To submit a URL to the article API make the following request :

    my $response = $client->article({ url => 'http://analyzethis.com' });

This simpler parameter syntax works too:

    my $response = $client->article('http://analyzethis.com');

To submit a URL to the classifier (aka analyze) API make the following request :

    my $response = $client->analyze({ url => 'http://analyzethis.com' });

Similarly, to use the other APIs make requests like this:

    $response = $client->frontpage({ url => 'http://analyzethis.com' });
    $response = $client->product({ url => 'http://analyzethis.com' });
    $response = $client->image({ url => 'http://analyzethis.com' });

If you want to extract key values out of the response, do so like this:

    print $response->{title};

If you want to see the full Diffbot response in all of it's glory do this:

    use Data::Dump qw(dump);
    print dump($response); #wowsa thats a lotta data!

If you want to control the timeout behaviour and the fields returned, make a request like this:

    my $response = $client->article({ 
        url => 'http://analyzethis.com',
        timeout => 30000, 
        fields => 'title,link,text'
    });

Here is an alternate syntax for making the same request:
    
    my $response = $client->query({
        request_type => 'article',
        query_args => { 
            url => 'http://analyzethis.com',
            timeout => 30000, 
            fields => 'title,link,text'
        }
    });

If something is going wrong and you want to get a better look at the full HTTP request and response make a request like this:
    
    use Data::Dump qw(dump)

    #$response is no the HTTP::Response object
    my $response = $client->query({
        request_type => 'article',
        query_args => { url => 'http://analyzethis.com', },
        return_http_response => 1 
    });

    print $response->status_line

## Submit content to Diffbot for analysis

If you want to submit local content to diffbot make a request like this:

    my $response = $client->article({ 
        url => 'http://analyzethis.com',
        content => $local_html_content
    });

This alternative syntax for submitting content to analyze works too:

    my $response = $client->query({
        request_type => 'article',
        query_args => { 
            url     => 'http://analyzethis.com',
            content => $local_html_content
        }
    });

Note that not all of the diffbot APIs support submitting local content for analysis. See the diffbot API docs for more details.

## Exceptions

See the built in docs for up to date exception listing and description. But given we are using Perl, it's easy! Just eval everything:

    my $response;
    eval {
        $response = $client->article({ url => 'http://analyzethis.com' });
    };
    if ($@) {
        print "Something went wrong: $@";
    }

## Can I havez some example code ?

The module ships with a fully functional script that uses the module, called query\_diffbot.pl
Run the following command to get up to date usage instructions:

    perl bin/query_diffbot.pl --help

-Initial commit by Zyle Zeeuwen-


Changes
Revision history for Perl extension Diffbot::Client.

0.01  Wed Jan  8 13:36:05 2014
	- original version authored by Kyle Zeeuwen; created by h2xs 1.23 with options
		-AX -b 5.12.0 Diffbot::Client



Makefile.PL
use 5.012000;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Diffbot::Client',
    VERSION_FROM      => 'lib/Diffbot/Client.pm', # finds $VERSION
    PREREQ_PM         => { # e.g., Module::Name => 1.1
        'Carp'              => 0,
        'Data::Dump'        => 0,
        'JSON'              => 0,
        'LWP::UserAgent'    => 0,
    },
    BUILD_REQUIRES    => {
        'Getopt::Long'      => 0,
        'Test::Deep'        => 0,
        'Test::Fake::HTTPD' => 0,
        'Test::More'        => 0,
        'Test::Exception'   => 0,
    },
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Diffbot/Client.pm', # retrieve abstract from module
       AUTHOR         => 'Kyle Zeeuwen <kyle.zeeuwen@gmail.com>') : ()
    ),
);


lib/Diffbot/Client.pm
package Diffbot::Client;

use 5.012000;
use strict;
use warnings;

our $VERSION = '0.01';

use Carp;
use JSON;
use LWP::UserAgent;

my $DEFAULTS = {
    'diffbot_host' => 'http://api.diffbot.com',
};

my $PATHS = {
    'article'   => 'v2/article',
    'frontpage' => 'v2/frontpage',
    'image'     => 'v2/image',
    'product'   => 'v2/product',
    'analyze'   => 'v2/classifier',
    'classifier' => 'v2/classifier',
};

sub new {
    my ($class, $args) = @_;

    my $self = bless {}, $class;

    croak "Missing token" unless $args->{token};
    $self->{token} = $args->{token};
    
    #get diffbot URL, verify starts with http:// and strip trailing slash
    $self->{diffbot_host} = $args->{diffbot_host} || $DEFAULTS->{diffbot_host};
    $self->{diffbot_host} = 'http://' . $self->{diffbot_host} unless
        $self->{diffbot_host} =~ qr{^http://};
    $self->{diffbot_host} =~ s{/$}{};
  
    eval {
        if ($args->{ua}) { 
            $self->{ua} = $args->{ua};
        }
        else {
            $self->{ua} = LWP::UserAgent->new();
        }
    };
    if ($@) { 
        croak "Error initializing WWW client: $@";
    }

    return $self;
}

#This dynamically generates sub routines:
for my $api (sort keys %$PATHS) {

    # turn off strict refs so that we can
    # register a method in the symbol table
    no strict "refs";

    *$api = sub {
        my ($self, $query_args) = @_;
        
        return $self->query({ 
            request_type => $api, 
            query_args   => $query_args
        });
    };
}

sub query {
    my ($self, $args) = @_;

    croak "Missing request_type" unless $args->{request_type};
    my $request_type = $args->{request_type};
    croak "Invalid request_type" unless $PATHS->{$request_type};

    #query args must be present. 
    #It can be a scalar (just the URL) or a hashref of args
    croak "Missing query_args" unless $args->{query_args};
    my $query_args = (ref($args->{query_args}) eq 'HASH')
        ? $args->{query_args}
        : { url => $args->{query_args} };

    #XXX: Additional URL validation here ?
    croak "Missing url" unless $query_args->{url};

    croak "Invalid timeout. Must be positive integer number of milliseconds" 
        if defined $query_args->{timeout} && $query_args->{timeout} !~ /^\d+$/;

    #XXX: No validation of fields,mode, or stats params
    
    my ($request_url, $http_request, $http_response);
    eval {

        #strip content from query args (or it will end up in URL)
        my $content = $query_args->{content};
        delete($query_args->{content});
        
        $request_url = $self->_build_get_url($request_type,$query_args);

        #if content is present, this is a POST, else it is a GET
        if ($content) {
            $http_request = HTTP::Request->new( POST => $request_url );
            $http_request->header( 'Accept' => 'application/json' );
            $http_request->header( 'Content-Type' => 'text/html' );
            
            #XXX: Must handle encoding of non-ascii here
            $http_request->content($content);

        }
        else {
            $http_request = HTTP::Request->new( GET => $request_url );
            $http_request->header( 'Accept' => 'application/json' );
        }
    };
    if ($@) {
        croak "Error creating HTTP Request for url '" . 
            $request_url || '' . "' : $@";
    }

    eval {
        $http_response = $self->{ua}->request($http_request);
    };
    if ($@) {
        croak "Error making HTTP Request for url '" . 
            $request_url || '' . "' : $@";
    }

    # for debugging / when the client seems to be misbehaving
    if ($args->{return_http_response}) {
        return $http_response;
    }

    my $diffbot_response;
    if ($http_response && $http_response->is_success) {
        my $json_string = $http_response->decoded_content;
        eval {
            $diffbot_response = decode_json($json_string);
        };
        if ($@) {
            croak "Could not parse the diffbot response for url '" . 
                $request_url || '' . "' : $@";
        }
    }
    else {
        croak "The diffbot server returned: " . $http_response->status_line . 
            " for url '" . $request_url || '' . "' : $@";
    }

    return $diffbot_response;
}

sub _build_get_url {
    my ($self, $request_type, $query_args) = @_;

    $query_args->{token} = $self->{token};

    my $diffbot_url = 
        $self->{diffbot_host} . '/' . $PATHS->{$request_type} .  
        '?' . join('&', map { "$_=$query_args->{$_}" } sort keys %$query_args );

    return $diffbot_url;
}

1;
__END__

=head1 NAME

Diffbot::Client - perl client for the diffbot REST API.

=head1 SYNOPSIS

  use Diffbot::Client;
  
  my $client = Diffbot::Client->new({ token => $token });
  
  my $diffbot_article_response   = $client->article({ url => $target_url });
  my $diffbot_frontpage_response = $client->frontpage({ url => $target_url });
  my $diffbot_product_response   = $client->product({ url => $target_url });
  my $diffbot_image_response     = $client->image({ url => $target_url });
  my $diffbot_classifier_response = $client->classifier({ url => $target_url });

  #note: analyze is an alias for classifier
  my $diffbot_analyze_response = $client->analyze({ url => $target_url });

=head1 DESCRIPTION

This module provides a Perl client for the diffbot REST API. Details of the REST API can be found here : http://www.diffbot.com/. You will need a token from diffbot to use their service.

There are five exposed APIs: article, frontpage, product, image, and classifier.
Each API can be queried in two ways:

=over

=item Fetch and Analyze 

You provide a URL, and diffbot fetches the URL and analyzes the content

=item Just Analyze 

You fetch the URL and provide the content, and diffbot analyzes the content

=back

In all cases, the return value is a hashref containing the response. See the API documentation for the specific fields returned.

=head1 Constructor

$client = Diffbot::Client->new($args)

=head2 $args values

$args is a hashref that contains the following key value pairs:

=over

=item * token: Required. Request a token from http://www.diffbot.com/

=item * diffbot_host: Optional. Change the diffbot_server host from the default http://api.diffbot.com

=item * ua: Optonal. You may pass in a LWP::UserAgent if you need to customize HTTP interaction. If not provided, a LWP::UserAgent will be created.  

=back

=head2 Exceptions

=over

=item * Missing token

You must provide a valid token

=item * Error initializing WWW client:

This is an inernal error that was raised while creating the LWP::UserAgent. This indicates a missing dependancy or networking issue.

=back

=head1 Fetch and Analyze Functions

=over

=item $response = $client->article($args);

=item $response = $client->frontpage($args);

=item $response = $client->product($args);

=item $response = $client->image($args);

=item $response = $client->image($args);

=item $response = $client->classifier($args);

=item $response = $client->analyze($args);

=back

=head2 $args values

$args is a hashref that contains the following key value pairs:

=over

=item * url: Required. The URL to fetch (must include http://)

=item * fields: Optional. Specify which fields to include in the response

=item * timeout: Optional. The number of milliseconds before diffbot will terminate the response

=back

=head2 $response

The response is a hash reference containing keys. See the API doc for the specific keys returned. As an example to access the 'links' key or the response:

$links = $response->{links};

=head2 Exceptions

=over

=item * Missing query_args

You did not pass an $args variable to the sub routine

=item * Missing url

You did not include a 'url' key value pair in $args

=item * Invalid timeout. 

You passed an invalid timeout value to the sub routine. The timeout must be positive integer number of milliseconds

=item * Error creating HTTP Request for url / Error making HTTP Request for url

This is an internal error. Your URL is likely malormed or your environment is missing Perl dependancies.

=item * Could not parse the diffbot response for url

The diffbot server response was not well formed JSON.

=item * The diffbot server returned: 

The diffbot server did not respond with a HTTP 200 OK

=back

=head1 Just Analyze Functions

Not Implemented

=head1 Advanced / Diagnostic Functions

=head2 $repsone=query($args)

=head3 Example

    $response = query({
       request_type => 'article',
       query_args   => {
           url     => 'http://analyzethis.com',
           timeout => 30
       },
       return_http_response => 1 
    }); 

If you need to see what is happenning under the hood this function is useful. this is the internal function that actually does all the work. Note this is not 'public' so the author does not provide any backward compatability in the future if you choose to use this sub routine

=head3 $args values

=over

=item * request_type: Required. The API to use : (i.e., article, frontpage, product, image, analyze, classifier)

=item * query_args: This is the hashref documented above (url, timeout) in the Fetch and Analyze section.

=item * return_http_response: Optional. If set to true, the sub routine will return the full HTTP::Repsonse instead of just the diffbot response. The reponse also contains the HTTP::Request. Useful for debugging.

=back

=head3 Additional Exceptions

=over

=item * Missing request_type / Invalid request_type

You must provide a request_type key with a valid API name (i.e., article, frontpage, product, image, analyze, classifier)

=back

=head1 SEE ALSO

Refer to the documentation of the API on http://www.diffbot.com/.

=head1 AUTHOR

Kyle Zeeuwen, <lt>kyle.zeeuwen@gmailcom.com<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014 by Diffbot

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.16.2 or,
at your option, any later version of Perl 5 you may have available.

=cut
